language/diem-framework/modules/DiemId.move

```diff
@@ -0,0 +1,144 @@
+    const EDIEMIDDOMAINMANAGER: u64 = 1;
+
+    /// Publish a `DiemIdDomains` resource under `created` with an empty `domains`.
+    /// Before sending or receiving any payments using Diem IDs, the Treasury Compliance account must send
+    /// a transaction that invokes `add_domain_id` to set the `domains` field with a valid domain
+    public fun publish_diem_id_domains(
+        created: &signer,
+    ) {
+        Roles::assert_parent_vasp_role(created);
+        assert(
+            !exists<DiemIdDomains>(Signer::address_of(created)),
+            Errors::already_published(EDIEMIDDOMAIN)
+        );
+        move_to(created, DiemIdDomains {
+            domains: Vector::empty(),
+        })
+    }
+    spec fun publish_diem_id_domains {
+        include Roles::AbortsIfNotParentVasp{account: created};
+        aborts_if has_diem_id_domains(Signer::spec_address_of(created)) with Errors::ALREADY_PUBLISHED;
+    }
+
+    public fun has_diem_id_domains(addr: address): bool {
+        exists<DiemIdDomains>(addr)
+    }
+
+    /// Publish a `DiemIdDomainManager` resource under `created` with an empty `diem_id_domain_events`.
+    /// When Treasury Compliance account sends a transaction that invokes `update_diem_id_domain`,
+    /// a `DiemIdDomainEvent` is emitted and added to `diem_id_domain_events`.
+    public fun publish_diem_id_domain_manager(
+        created: &signer,
+    ) {
+        Roles::assert_treasury_compliance(created);
+        assert(
+            !exists<DiemIdDomainManager>(Signer::address_of(created)),
+            Errors::already_published(EDIEMIDDOMAINMANAGER)
+        );
+        move_to(
+            created,
+            DiemIdDomainManager {
+                diem_id_domain_events: Event::new_event_handle<DiemIdDomainEvent>(created),
+            }
+        );
+    }
+
+    spec define spec_has_diem_id_domain_manager(addr: address): bool {
+        exists<DiemIdDomainManager>(addr)
+    }
+
+    spec fun publish_diem_id_domain_manager {
+        include Roles::AbortsIfNotTreasuryCompliance{account: created};
+        aborts_if spec_has_diem_id_domain_manager(Signer::spec_address_of(created)) with Errors::ALREADY_PUBLISHED;
+    }
+
+    public fun update_diem_id_domain(
+        tc_account: &signer,
+        to_update_address: address,
+        domain: vector<u8>,
+        is_remove: bool
+    ) acquires DiemIdDomainManager, DiemIdDomains {
+        Roles::assert_treasury_compliance(tc_account);
+        let account_domains = borrow_global_mut<DiemIdDomains>(to_update_address);
+        let diem_id_domain = DiemIdDomain {domain: domain};
+        if (is_remove) {
+            let (has, index) = Vector::index_of(&account_domains.domains, &diem_id_domain);
+            if (has) {
+                Vector::remove(&mut account_domains.domains, index);
+            };
+        } else {
+            Vector::push_back(&mut account_domains.domains, copy diem_id_domain);
```

Comment:
We should always do a check for if the domain already exists, we don't want to allow for adding a duplicate.